### PR TITLE
Feat/3905  avatar update

### PR DESCRIPTION
**Pull request by grignards**
#3905 

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

File: frontend/i18next.scanner/pt/translation.json [hunk 1]
```diff
@@ -335,5 +332,9 @@
   "Loading": "Carregando",
   "Information_plural": "Informações",
   "Personal page": "Página pessoal",
-  "My profile": "Meu perfil"
+  "My profile": "Meu perfil",
+  "Upload an image": "Carregar imagem",
+  "Recommended dimensions:": "Dimensões recomendadas:",
+  "Maximum size: {{size}}Mb": "Tamanho máximo: {{size}}Mb",
```
Comment (gbonaspetti) [inline frontend/i18next.scanner/pt/translation.json:338]:
```suggestion
  "Maximum size: {{size}} Mb": "Tamanho máximo: {{size}} Mb",
```

File: frontend/src/container/GuestUpload.jsx [hunk 2]
```diff
@@ -137,16 +141,16 @@ class GuestUpload extends React.Component {
     }
 
     this.setState(previousState => ({
-      fileToUploadList: [
-        ...previousState.fileToUploadList,
-        ...newFileList.filter(newFile => !state.fileToUploadList.some(stateFile => stateFile.name === newFile.name))
+      fileUploadList: [
+        ...previousState.fileUploadList,
+        ...newFileUploadList
```
Comment (raphj) [inline frontend/src/container/GuestUpload.jsx:146]: Does this account for duplicates?

File: frontend/src/component/PublicProfile/ProfileMainBar.jsx [hunk 3]
```diff
@@ -1,27 +1,45 @@
...
+        />
+        <Avatar
+          customClass='profile__mainBar__avatar__medium'
+          user={hasUser ? props.displayedUser : {}}
+          size={AVATAR_SIZE.MEDIUM}
+          filenameInUrl={props.avatarFilenameInUrl}
+        />
+        {props.changeAvatarEnabled && (
+          <IconButton
+            text='' title={props.t('Change avatar')}
```
Comment (gbonaspetti) [inline frontend/src/component/PublicProfile/ProfileMainBar.jsx:34]:
```suggestion
            text='' 
            title={props.t('Change avatar')}
```

`text` is required?

File: frontend/src/container/GuestUpload.jsx [hunk 4]
```diff
@@ -181,9 +185,11 @@ class GuestUpload extends React.Component {
     if (!this.validateForm()) return false
 
     const formData = new FormData()
-
-    state.fileToUploadList.forEach((uploadFile, index) => {
-      formData.append(`file_${index}`, uploadFile)
+    // TODO - SG - 2021-01-22 - We really should use the uploadFile() function here
+    // But first check if guest-upload backend could be compatible to one FormData field for all files ('files)
```
Comment (raphj) [inline frontend/src/container/GuestUpload.jsx:189]: Missing quote?
```

File: frontend/src/container/PublicProfile.jsx [hunk 5]
```diff
@@ -62,46 +80,66 @@ export class PublicProfile extends React.Component {
...
+  onChangeAvatarSuccess = () => this.props.dispatch(updateUserProfileAvatarName(`avatar-${Date.now()}`))
+
+  onCloseUploadPopup = () => this.setState({ displayUploadPopup: undefined })
+
   render () {
     const { props, state } = this
-
+    if (!state.displayedUser) return null
+    const changeImageEnabled = (state.displayedUser.userId === props.user.userId) || props.user.profile === PROFILE.ADMINISTRATOR
+    //const uploadAvatarUrl = `${FETCH_CONFIG.apiUrl}/users/${state.displayedUser.userId}/avatar/raw`
```
Comment (gbonaspetti) [inline frontend/src/container/PublicProfile.jsx:123]: Forgotten comment?

File: frontend/src/container/PublicProfile.jsx [hunk 6]
```diff
@@ -62,46 +80,66 @@ export class PublicProfile extends React.Component {
...
+  onChangeAvatarClick = () => this.setState({ displayUploadPopup: POPUP_DISPLAY_STATE.AVATAR })
+
+  onChangeAvatarSuccess = () => this.props.dispatch(updateUserProfileAvatarName(`avatar-${Date.now()}`))
+
+  onCloseUploadPopup = () => this.setState({ displayUploadPopup: undefined })
+
   render () {
     const { props, state } = this
-
+    if (!state.displayedUser) return null
```
Comment (gbonaspetti) [inline frontend/src/container/PublicProfile.jsx:121]: This line does not block the loading screen while the API has not yet given the answer to the displayedUser?

File: frontend/src/container/PublicProfile.jsx [hunk 7]
```diff
@@ -62,46 +80,66 @@ export class PublicProfile extends React.Component {
...
+    const changeImageEnabled = (state.displayedUser.userId === props.user.userId) || props.user.profile === PROFILE.ADMINISTRATOR
+    //const uploadAvatarUrl = `${FETCH_CONFIG.apiUrl}/users/${state.displayedUser.userId}/avatar/raw`
+    const uploadAvatarUrl = `${FETCH_CONFIG.apiUrl}/workspaces/1/files`
```
Comment (gbonaspetti) [inline frontend/src/container/PublicProfile.jsx:124]: `workspaces/1`?

File: frontend_app_file/src/container/PopupCreateFile.jsx [hunk 8]
```diff
@@ -42,6 +34,7 @@ class PopupCreateFile extends React.Component {
     ])
   }
 
+  // File specific ??
```
Comment (gbonaspetti) [inline frontend_app_file/src/container/PopupCreateFile.jsx:37]: Forgotten comment?

File: frontend_app_file/src/container/PopupCreateFile.jsx [hunk 9]
```diff
@@ -52,6 +45,7 @@ class PopupCreateFile extends React.Component {
     this.setHeadTitle()
   }
 
+  // File specific
```
Comment (gbonaspetti) [inline frontend_app_file/src/container/PopupCreateFile.jsx:48]: Forgotten comment?

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 10]
```diff
@@ -122,5 +123,12 @@
   "Edit": "Editar",
   "Validate anyway": "Validar de qualquer forma",
   "Home": "Início",
-  "Error while getting breadcrumbs": "Erro ao obtenir a navegação estrutural"
+  "Only one file is allowed": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:126]:
```suggestion
  "Only one file is allowed": "Só é permitido um ficheiro",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 11]
```diff
@@ -122,5 +123,12 @@
...
+  "Files already added:": "Ficheiros já subidos:",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:127]:
```suggestion
  "Files already added:": "Ficheiros já adicionados:",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 12]
```diff
@@ -122,5 +123,12 @@
...
+  "The type of this file is not allowed": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:128]:
```suggestion
  "The type of this file is not allowed": "O tipo deste ficheiro não é permitido",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 13]
```diff
@@ -122,5 +123,12 @@
...
+  "The file is too big": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:129]:
```suggestion
  "The file is too big": "O ficheiro é demasiado grande",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 14]
```diff
@@ -122,5 +123,12 @@
...
+  "Error while uploading file": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:130]:
```suggestion
  "Error while uploading file": "Erro ao carregar o ficheiro",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 15]
```diff
@@ -122,5 +123,12 @@
...
+  "Error while uploading file(s)": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:131]:
```suggestion
  "Error while uploading file(s)": "Erro ao carregar o(s) ficheiro(s)",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 16]
```diff
@@ -122,5 +123,12 @@
...
+  "Please wait until the upload ends": "__NOT_TRANSLATED__",
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:132]:
```suggestion
  "Please wait until the upload ends": "Por favor aguarde até que o carregamento termine",
```

File: frontend_lib/i18next.scanner/pt/translation.json [hunk 17]
```diff
@@ -122,5 +123,12 @@
...
+  "Avatar of {{publicName}}": "__NOT_TRANSLATED__"
```
Comment (gbonaspetti) [inline frontend_lib/i18next.scanner/pt/translation.json:133]:
```suggestion
  "Avatar of {{publicName}}": "Avatar de {{publicName}}"
```

File: frontend_lib/src/component/Avatar/Avatar.jsx [hunk 18]
```diff
@@ -1,8 +1,9 @@
 import React from 'react'
 import PropTypes from 'prop-types'
 import classnames from 'classnames'
+import { translate } from 'react-i18next'
 
-// require('./Avatar.styl') // see https://github.com/tracim/tracim/issues/1156
+require('./Avatar.styl')
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Avatar/Avatar.jsx:6]: Why this is now required?

File: frontend_lib/src/component/Avatar/Avatar.jsx [hunk 19]
```diff
@@ -45,44 +46,58 @@ export class Avatar extends React.Component {
   render () {
     const { props } = this
 
-    const generatedColor = this.generateColorFromName(props.publicName)
+    const publicName = props.user.publicName || props.user.public_name
+    const hasAvatar = Object.keys(props.user).includes('hasAvatar') ? props.user.hasAvatar : props.user.has_avatar
+    const letterAvatar = publicName ? this.getTwoLetters(publicName.toUpperCase()) : '?'
+    const sizeAsNumber = parseInt(props.size.replace('px', ''))
+
+    const generatedColor = publicName ? this.generateColorFromName(publicName) : '#000'
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Avatar/Avatar.jsx:54]: We have the same check at `generateColorFromName`, I don't mind that it is done here, but please use the same default color.

File: frontend_lib/src/component/Avatar/Avatar.jsx [hunk 20]
```diff
@@ -45,44 +46,58 @@ export class Avatar extends React.Component {
...
-            borderRadius: props.size,
             backgroundColor: generatedColor,
             fontSize: fontSize
           }}
         >
-          {this.getTwoLetter(props.publicName.toUpperCase())}
+          {hasAvatar ? (
+            <img
+              className='avatar__img'
+              src={`${props.apiUrl}/workspaces/1/files/1/preview/jpg/${sizeAsNumber}x${sizeAsNumber}/${props.filenameInUrl}`}
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Avatar/Avatar.jsx:76]: Is it normal that it is "fixed" in the space 1 file 1?

File: frontend_lib/src/component/Avatar/Avatar.styl [hunk 21]
```diff
@@ -1,7 +1,14 @@
+@import "../../css/Variable.styl"
+
 .avatar
   display flex
   justify-content center
   align-items center
   text-align center
   font-weight bold
   color #fdfdfd
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Avatar/Avatar.styl:9]:
```suggestion
  color offWhite
```

File: frontend_lib/src/component/Button/IconButton.styl [hunk 22]
```diff
@@ -46,3 +49,18 @@
     padding 0 5px
     &:hover
       text-decoration underline
+
+  &__pins
+    padding: 0
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Button/IconButton.styl:54]:
```suggestion
    padding 0
```

File: frontend_lib/src/component/Button/IconButton.styl [hunk 23]
```diff
@@ -46,3 +49,18 @@
...
+    border-radius: 50% 50%
```
Comment (gbonaspetti) [inline frontend_lib/src/component/Button/IconButton.styl:55]:
```suggestion
    border-radius 50% 50%
```

File: functionnal_tests/cypress/integration/profile/profile_images_update.spec.js [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+  for (username of allowedUsernames) {
+    describe(`As ${username}`, () => {
+      before(function () {
+        cy.resetDB()
+        cy.setupBaseDB()
+        cy.loginAs(username)
+        cy.visitPage({ pageName: PAGES.PROFILE, params: { userId: baseUser.user_id } })
+      })
+
+      it("should have an avatar button which allows to change the avatar", () => {
```
Comment (gbonaspetti) [inline functionnal_tests/cypress/integration/profile/profile_images_update.spec.js:18]:
```suggestion
      it('should have an avatar button which allows to change the avatar', () => {
```

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

File: frontend/src/container/GuestUpload.jsx (hunk 2 above)
Comment (grignards) [reply frontend/src/container/GuestUpload.jsx:146]: The code above takes care of removing the duplicates (and warns the user if there is one).

File: frontend_lib/src/component/Avatar/Avatar.jsx (hunk 20 above)
Comment (grignards) [reply frontend_lib/src/component/Avatar/Avatar.jsx:76]: No, I added this to test without the backend code, I'll change it to reflect the correct URL.

File: frontend/src/component/PublicProfile/ProfileMainBar.jsx (hunk 3 above)
Comment (grignards) [reply frontend/src/component/PublicProfile/ProfileMainBar.jsx:34]: It was but isn't anymore, I'll remove it.

File: frontend/src/container/PublicProfile.jsx (hunk 5 above)
Comment (grignards) [reply frontend/src/container/PublicProfile.jsx:123]: Same reason as above, I will change to the good url.

File: frontend/src/container/PublicProfile.jsx (hunk 6 above)
Comment (grignards) [reply frontend/src/container/PublicProfile.jsx:121]: Yes, I'll remove it (I added it for starting the development).

File: frontend_lib/src/component/Avatar/Avatar.jsx (hunk 18 above)
Comment (grignards) [reply frontend_lib/src/component/Avatar/Avatar.jsx:6]: I added  the styling for the image (border-radius) inside.
I read the linked issue, it isn't valid anymore as we already use `require()` of stylus files in many frontend_lib components.

File: frontend/src/component/PublicProfile/ProfileMainBar.jsx [hunk 25]
```diff
@@ -1,27 +1,46 @@
...
+        <Avatar
+          customClass='profile__mainBar__avatar__medium'
+          user={hasUser ? props.displayedUser : {}}
+          size={AVATAR_SIZE.MEDIUM}
+          filenameInUrl={props.avatarFilenameInUrl}
+        />
+        {props.changeAvatarEnabled && (
+          <IconButton
+            text='' 
+            title={props.t('Change avatar')}
```
Comment (gbonaspetti) [inline frontend/src/component/PublicProfile/ProfileMainBar.jsx:35]:
```suggestion
          <IconButton
            title={props.t('Change avatar')}
```

File: frontend/src/container/PublicProfile.jsx [hunk 26]
```diff
@@ -62,46 +81,64 @@ export class PublicProfile extends React.Component {
...
+              uploadUrl={`${avatarBaseUrl}/avatar/raw`}
+              httpMethod='PUT'
+              color={GLOBAL_primaryColor}
+              handleClose={this.onCloseUploadPopup}
+              handleSuccess={this.onChangeImageSuccess}
+              allowedMimeTypes={ALLOWED_IMAGE_MIMETYPES}
+              maximumFileSize={MAXIMUM_AVATAR_SIZE}
+            >
+              <i className='fa fa-fw fa-arrows-alt' /> {props.t('Recommended dimensions:')} 100x100px<br />
+              <i className='fa fa-fw fa-image' /> {props.t('Maximum size: {{size}}Mb', { size: MAXIMUM_AVATAR_SIZE / (1024 * 1024) })}
```
Comment (gbonaspetti) [inline frontend/src/container/PublicProfile.jsx:139]:
```suggestion
              <i className='fa fa-fw fa-image' /> {props.t('Maximum size: {{size}} Mb', { size: MAXIMUM_AVATAR_SIZE / (1024 * 1024) })}
```

File: frontend_lib/src/helper.js [hunk 27]
```diff
@@ -697,3 +697,5 @@ export const buildContentPathBreadcrumbs = async (apiUrl, content) => {
       throw new Error('Error getting breadcrumbs data')
   }
 }
+
+export const getAvatarBaseUrl = (apiUrl, userId) =>  `${props.apiUrl}/users/${}/avatar`
```
Comment (gbonaspetti) [inline frontend_lib/src/helper.js:701]: The `${}` is normal?

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

File: frontend_lib/src/helper.js (hunk 27 above)
Comment (raphj) [reply frontend_lib/src/helper.js:701]:
```suggestion
export const getAvatarBaseUrl = (apiUrl, userId) =>  `${props.apiUrl}/users/${userId}/avatar`
```

File: frontend/src/container/PublicProfile.jsx (hunk 26 above)
Comment (raphj) [reply frontend/src/container/PublicProfile.jsx:139]: Even this:
```suggestion
              <i className='fa fa-fw fa-image' /> {props.t('Maximum size: {{size}} MB', { size: MAXIMUM_AVATAR_SIZE / (1024 * 1024) })}
```

File: frontend/src/container/GuestUpload.jsx [hunk 28]
```diff
@@ -114,17 +116,19 @@ class GuestUpload extends React.Component {
...
+    const alreadyPresentList = droppedFileUploadList.filter(newFileUpload => isFileUploadInList(newFileUpload, state.fileUploadList))
+    const newFileUploadList = droppedFileUploadList.filter(newFileUpload => !isFileUploadInList(newFileUpload, state.fileUploadList))
+    if (alreadyPresentList.length) {
       GLOBAL_dispatchEvent({
         type: CUSTOM_EVENT.ADD_FLASH_MSG,
         data: {
           msg: (
             <div>
-              {props.t('Files already uploaded:')}
+              {props.t('Files already added:')}
```
Comment (raphj) [inline frontend/src/container/GuestUpload.jsx:128]: Well, technically, this would be "A file with the same name is already there", since the user could try to upload two different files with the same file name (it already happened to me).
But this is out of scope of this PR probably

File: frontend/src/container/GuestUpload.jsx (hunk 2 above)
Comment (raphj) [reply frontend/src/container/GuestUpload.jsx:146]: Ah, right, so maybe don't change the state if  `alreadyPresentList.length`

File: frontend/src/container/PublicProfile.jsx [hunk 29]
```diff
@@ -62,46 +81,64 @@ export class PublicProfile extends React.Component {
...
+              uploadUrl={`${avatarBaseUrl}/avatar/raw`}
+              httpMethod='PUT'
+              color={GLOBAL_primaryColor}
+              handleClose={this.onCloseUploadPopup}
+              handleSuccess={this.onChangeImageSuccess}
+              allowedMimeTypes={ALLOWED_IMAGE_MIMETYPES}
+              maximumFileSize={MAXIMUM_AVATAR_SIZE}
+            >
+              <i className='fa fa-fw fa-arrows-alt' /> {props.t('Recommended dimensions:')} 100x100px<br />
+              <i className='fa fa-fw fa-image' /> {props.t('Maximum size: {{size}} MB', { size: MAXIMUM_AVATAR_SIZE / (1024 * 1024) })}
```
Comment (gbonaspetti) [inline frontend/src/container/PublicProfile.jsx:139]: The keys are case-sensitive, if you use `MB` here, don't forget to change at the translation files.

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

File: functionnal_tests/cypress/integration/profile/profile_images_update.spec.js [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+      cy.setupBaseDB()
+      cy.login(defaultAdmin)
+      cy.createRandomUser().then(user => {
+        cy.addUserToWorkspace(user.user_id, baseWorkspace.workspace_id)
+        cy.login(baseUser)
+        cy.visitPage({ pageName: PAGES.PROFILE, params: { userId: user.user_id } })
+      })
+    })
+
+    it("should NOT have a change avatar button", () => {
```
Comment (gbonaspetti) [inline functionnal_tests/cypress/integration/profile/profile_images_update.spec.js:44]:
```suggestion
    it('should NOT have a change avatar button', () => {
```

**Review by gbonaspetti — CHANGES_REQUESTED** [review]

**Comment by gbonaspetti** [comment]
During tests:

- The button is not round at Chrome
- If I'm logged as admin and I change another user's avatar I have the error `{"message": "'NoneType' object has no attribute 'filename'", "details": {"error_detail": {}}, "code": null}`
- The avatar is not update at activity feed (personal and for space), member's list nor search results
- Error at "Requests to join the space" on advanced dashboard

File: frontend/src/container/GuestUpload.jsx [hunk 31]
```diff
@@ -181,9 +185,11 @@ class GuestUpload extends React.Component {
     if (!this.validateForm()) return false
 
     const formData = new FormData()
-
-    state.fileToUploadList.forEach((uploadFile, index) => {
-      formData.append(`file_${index}`, uploadFile)
+    // TODO - SG - 2021-01-22 - We really should use the uploadFile() function here
```
Comment (Skylsmoi) [inline frontend/src/container/GuestUpload.jsx:188]: Please create an issue about this and copy the link in the comment

**Review by gbonaspetti — APPROVED** [review]

File: frontend/src/container/PublicProfile.jsx [hunk 32]
```diff
@@ -62,46 +81,83 @@ export class PublicProfile extends React.Component {
...
 
     const fetchGetUser = await props.dispatch(getAboutUser(userId))
+    if (this.handleFetchErrors(
+      fetchGetUser,
+      [{ status: 400, code: 1001, message: props.t('Unknown user') }],
+      props.t('Error while loading user')
+    )) return
+
+    const apiUser = { ...serialize(fetchGetUser.json, serializeUserProps) }
+    this.setState(oldState => {
```
Comment (Skylsmoi) [inline frontend/src/container/PublicProfile.jsx:107]: the convention is to name the parameter `prevState` or `previousState`

**Review by Skylsmoi — COMMENTED** [review]
be advice that this PR will have many merge conflicts with https://github.com/tracim/tracim/pull/4086

**Comment by gbonaspetti** [comment]
Tested on branch.

File: frontend/src/container/PublicProfile.jsx (hunk 32 above)
Comment (grignards) [reply frontend/src/container/PublicProfile.jsx:107]: If it's ok with you I'll address this naming issue with #4088.

File: frontend/src/container/GuestUpload.jsx (hunk 31 above)
Comment (grignards) [reply frontend/src/container/GuestUpload.jsx:188]: Issue created (https://github.com/tracim/tracim/issues/4090), I'll add the comment in #4088.

**Comment by asf-algoo** [comment]
Closing this PR to re-open (and merge) with asf's user.

**Review by Skylsmoi — APPROVED** [review]

File: frontend/src/container/PublicProfile.jsx (hunk 32 above)
Comment (Skylsmoi) [reply frontend/src/container/PublicProfile.jsx:107]: sure
